src/Charcoal/Source/Database/DatabaseFilter.php

```diff
@@ -226,6 +226,45 @@ protected function byPredicate()
                     $conditions[] = sprintf('%1$s %2$s (\'%3$s\')', $target, $operator, $value);
                     break;
 
+                case 'BETWEEN':
+                case 'NOT BETWEEN':
+                    if (!is_array($value) || (is_array($value) && count($value) < 2)) {
+                        throw new UnexpectedValueException(sprintf(
+                            'Array is required as value on field "%s" for "%s"',
+                            $target,
+                            $operator,
+                        ));
+                    }
+
+                    $fromValue = $value[0];
+                    $toValue   = end($value);
+
+                    if (empty($fromValue) || empty($toValue)) {
+                        throw new UnexpectedValueException(sprintf(
+                            'Two values are required on field "%s" for "%s"',
+                            $target,
+                            $operator,
+                        ));
+                    }
+
+                    // Check if querying dates
+                    try {
+                        new \DateTime($fromValue);
+                        new \DateTime($toValue);
+                        $isDate = true;
+                    } catch (\Exception $e) {
+                        $isDate = false;
+                    }
+
+                    $conditions[] = sprintf(
+                        '%1$s %2$s %3$s AND %4$s',
+                        $target,
+                        $operator,
+                        $isDate ? 'CAST(\''.$fromValue.'\' AS DATE)' : '\''.$fromValue.'\'',
+                        $isDate ? 'CAST(\''.$toValue.'\' AS DATE)' : '\''.$toValue.'\'',
+                    );
```

Comment:
The following is more readable.

```suggestion
                    if ($isDate) {
                        $fromExpr = 'CAST(\''.$fromValue.'\' AS DATE)';
                        $toExpr   = 'CAST(\''.$toValue.'\' AS DATE)';
                    } else {
                        $fromExpr = '\''.$fromValue.'\'';
                        $toExpr   = '\''.$toValue.'\'';
                    }

                    $conditions[] = sprintf(
                        '%1$s %2$s %3$s AND %4$s',
                        $target,
                        $operator,
                        $fromExpr,
                        $toExpr
                    );
```